### PR TITLE
ORC-852: Allow DynamicByteArray to return a ByteBuffer

### PR DESCRIPTION
**Pull request by belugabehr**
### What changes were proposed in this pull request?
Allow `DyanmicByteArray` and `Dictionary` to return a ByteBuffer of its content.

### Why are the changes needed?
Performance.

### How was this patch tested?
No change to functionality. Existing unit tests used.

### CODE REVIEW

File: java/core/src/java/org/apache/orc/impl/DynamicByteArray.java [hunk 1]
```diff
@@ -294,6 +295,16 @@ public void setByteBuffer(ByteBuffer result, int offset, int length) {
     return result;
   }
 
+  public ByteBuffer get(int offset, int length) {
+    final int currentChunk = offset / chunkSize;
+    final int currentOffset = offset % chunkSize;
+    final int currentLength = Math.min(length, chunkSize - currentOffset);
+    if (currentLength == length) {
+      return ByteBuffer.wrap(data[currentChunk], currentOffset, length);
```
Comment (belugabehr) [inline java/core/src/java/org/apache/orc/impl/DynamicByteArray.java:303]: This is the "win" right here.  If the data requested is contiguous within a single data chunk, then a view to the data can be returned without needing to copy into a buffer.

**Comment by belugabehr** [comment]
Using ByteBuffer here will allow us to remove the `Text` version of this method in a subsequent release and move the project closer to removing Hadoop as a dependency (ORC-508)

**Comment by pgaref** [comment]
@belugabehr can we resolve conflicts here please?

**Comment by belugabehr** [comment]
@pgaref Will do, however, I think it would be better to get #735, #748 in first since they remove some instances of `Text` and then I can re-work this one to convert the remaining pieces to `ByteBuffer`

File: java/core/src/java/org/apache/orc/impl/StringHashTableDictionary.java [hunk 2]
```diff
@@ -120,6 +122,11 @@ public void getText(Text result, int positionInKeyOffset) {
     DictionaryUtils.getTextInternal(result, positionInKeyOffset, this.keyOffsets, this.byteArray);
   }
 
+  @Override
+  public ByteBuffer getText(int positionInKeyOffset) {
+    return DictionaryUtils.getTextInternal(positionInKeyOffset, this.keyOffsets, this.byteArray);
+  }
```
Comment (kbendick) [inline java/core/src/java/org/apache/orc/impl/StringHashTableDictionary.java:128]: Is it safe to be exporting this ByteBuffer? If it is reused / overwritten in some way, will that affect any of the existing code.

Specifically, I'm wondering about the now returned `ByteBuffer` from `setByteBuffer`, as the data can be returned without any copying.

Comment (belugabehr) [reply java/core/src/java/org/apache/orc/impl/StringHashTableDictionary.java:128]: Hey @kbendick,

> Is it safe to be exporting this ByteBuffer? If it is reused / overwritten in some way, will that affect any of the existing code.

I guess everyone has a different scale of what "safe" is.  If your concern is regarding the callers changing values in the returned ByteBuffer affected the data stored in the Dictionary, then yes, that is a possibility right now.  There are two ways to handle that:

1. Document this behavior and assume the caller will handle the data responsibly
2. Return a read-only `ByteBuffer` which will prevent callers from modifying the data

I'm not sure I understand the second part of your question.

Comment (belugabehr) [reply java/core/src/java/org/apache/orc/impl/StringHashTableDictionary.java:128]: On the second part, if you are referring to the new value returned by `setByteBuffer`, there's no functionality change there.  I am simply returning the passed in value for convenience:

```java
ByteBuffer bb = ByteBuffer.allocate(16);
dictionary.setByteBuffer(bb, 0, 16);
System.out.println(bb);

// v.s.

ByteBuffer bb = ByteBuffer.allocate(16);
System.out.println(dictionary.setByteBuffer(bb, 0, 16));
```

Comment (kbendick) [reply java/core/src/java/org/apache/orc/impl/StringHashTableDictionary.java:128]: That was indeed what I was referring to. That answers my question. Thank you @belugabehr!

**Comment by belugabehr** [comment]
@pgaref @kbendick Got the other items merged into the project, so this one is now ready.  Thanks!

**Comment by belugabehr** [comment]
@pgaref @kbendick Care to take a look over once more? Thank you!

File: java/core/src/java/org/apache/orc/impl/DictionaryUtils.java [hunk 3]
```diff
@@ -46,6 +47,18 @@ public static void getTextInternal(Text result, int position,
     byteArray.setText(result, offset, length);
   }
 
+  public static ByteBuffer getTextInternal(int position, DynamicIntArray keyOffsets,
```
Comment (pgaref) [inline java/core/src/java/org/apache/orc/impl/DictionaryUtils.java:50]: Shall we name this explicitly getTexBuffertInternal ?

File: java/core/src/java/org/apache/orc/impl/DictionaryUtils.java [hunk 4]
```diff
@@ -46,6 +47,18 @@ public static void getTextInternal(Text result, int position,
...
+      DynamicByteArray byteArray) {
+    final int offset = keyOffsets.get(position);
+    final int length;
+    if (position + 1 == keyOffsets.size()) {
+      length = byteArray.size() - offset;
+    } else {
+      length = keyOffsets.get(position + 1) - offset;
+    }
```
Comment (pgaref) [inline java/core/src/java/org/apache/orc/impl/DictionaryUtils.java:58]: Length calculation is repeated at least 5 times already -- lets move this to a helper function

File: java/core/src/java/org/apache/orc/impl/Dictionary.java [hunk 5]
```diff
@@ -52,6 +53,8 @@
    */
   void getText(Text result, int position);
 
+  ByteBuffer getText(int position);
```
Comment (pgaref) [inline java/core/src/java/org/apache/orc/impl/Dictionary.java:56]: maybe name getTextBuffer instead?

**Review by pgaref — COMMENTED** [review]
Hey @belugabehr  -- left some comments here, let me know what you think

File: java/core/src/java/org/apache/orc/impl/DictionaryUtils.java (hunk 3 above)
Comment (dongjoon-hyun) [inline java/core/src/java/org/apache/orc/impl/DictionaryUtils.java:50]: Could you add a function description please, @belugabehr ?

File: java/core/src/java/org/apache/orc/impl/DictionaryUtils.java (hunk 4 above)
Comment (dongjoon-hyun) [reply java/core/src/java/org/apache/orc/impl/DictionaryUtils.java:58]: I agree with @pgaref 's perspective. Could you elaborate a little bit more about your refactoring idea? Is that better?
>  I actually think this DictionaryUtils class can probably be removed and roll the functionality into StringHashTable directly.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM with one minor comment and one question. Thank you, @belugabehr .

File: java/core/src/java/org/apache/orc/impl/DictionaryUtils.java (hunk 3 above)
Comment (belugabehr) [reply java/core/src/java/org/apache/orc/impl/DictionaryUtils.java:59]: Done!

**Comment by belugabehr** [comment]
@dongjoon-hyun I made the change regarding JavaDocs.  Please review.

In regards to my comment:

> I actually think this DictionaryUtils class can probably be removed and roll the functionality into StringHashTable directly.

Looking into it a bit more, I think the best thing to do is to create a new data structure that stores the `DynamicByteArray` and its offsets (`DynamicIntArray`).  It should govern access and resizing the two as needed.  That would remove `DictionaryUtils`, make the methods more unit test-able and remove a bunch of logic from the `StringHashTableDictionary` class that deals with maintenance (like growing the arrays).  Alas, I still believe this should be accomplished as a separate issue/task.

**Comment by dongjoon-hyun** [comment]
Thank you, @belugabehr .
Merged to main.

**Pull request merged**